include/sys/util.h

```diff
@@ -471,7 +471,7 @@ char *utf8_lcpy(char *dst, const char *src, size_t n);
  *
  * @retval expr As a boolean return, if false then it has timed out.
  */
-#define wait_for(expr, timeout, delay_stmt)                                                        \
```

Comment:
While macros should be uppercase and I understand that this gives issues with CCP, I am not sure this is the right solution to the problem, nor the correct way of handling it. 

1) To me it seems like a sub-par design to have a function `wait_for` (from CPP) and a macro `WAIT_FOR` that kind of does the same thing, but not really

2) This should have been an inline function to begin with instead of a macro (see https://www.kernel.org/doc/html/v4.10/process/coding-style.html#macros-enums-and-rtl)

3) Renaming this is a stable API change, and should perhaps require an RFC (https://docs.zephyrproject.org/latest/reference/api/api_lifecycle.html?highlight=stable%20api#introducing-incompatible-changes) - Someone like @carlescufi can probably provide more insight here.

Reply:
> This should have been an inline function to begin with instead of a macro (see https://www.kernel.org/doc/html/v4.10/process/coding-style.html#macros-enums-and-rtl)

`expr` in this macro can be an expression. This cannot be emulated with a function call in standard C. The expression would have to be defined in a function, which can only be defined at top level in the file, with no access to local variables.

Reply:
> Renaming this is a stable API change

Yes. We should also have a deprecation period. We can define both names in C environments for a few releases.

Reply:
Actually, this macro is completely new. So we don't need a deprecation period. It was introduced in https://github.com/zephyrproject-rtos/zephyr/commit/3ab10f2d52a5f802e6dc8e2472a8563e70de69fc.

Reply:
Thanks for the great insight @alwa-nordic  :D 

This should be an easy change then, and my comments can be considered resolved.

Reply:
`wait_for` is implemented using statement expressions. This is a GNU C feature. Is the use of GNU C features allowed in Zephyr?

Reply:
GNU C features does given some issues if building with e.g. `native_posix` which is using `c99`

Reply:
> `wait_for` is implemented using statement expressions. This is a GNU C feature. Is the use of GNU C features allowed in Zephyr?

@alwa-nordic Good point. That really depends on how "de facto standard" that particular extension is. At least, statement expressions are currently supported by gcc and clang; but, other proprietary compilers (notoriously, IAR) may not, and that is going to be a problem for the project since one of our goals is to support many different toolchains.

In general, we do not want to introduce any new features that depend on the GNU extensions. In that sense, I would actually suggest to revert 3ab10f2d52a5f802e6dc8e2472a8563e70de69fc and possibly re-introduce it without using the GNU extensions.

cc @teburd @peter-mitsis @andyross @dcpleung @nashif

Reply:
> 3\. Renaming this is a stable API change, and should perhaps require an RFC (https://docs.zephyrproject.org/latest/reference/api/api_lifecycle.html?highlight=stable%20api#introducing-incompatible-changes) - Someone like @carlescufi can probably provide more insight here.

Yes, given that this macro is used everywhere I'd suggest following the Stable API Change procedure.

Reply:
> Yes, given that this macro is used everywhere I'd suggest following the Stable API Change procedure.

@carlescufi It is actually not used anywhere except in the test for itself. This was just recently introduced (see https://github.com/zephyrproject-rtos/zephyr/commit/3ab10f2d52a5f802e6dc8e2472a8563e70de69fc) and is not part of any releases.

Reply:
An inline function would be difficult here as it would require all the conditions to be function pointers. The macro makes it easier to write the condition being waited for.